cocotb/runner.py

```diff
@@ -0,0 +1,800 @@
+
+    def set_env(self) -> None:
+
+        for e in os.environ:
+            self.env[e] = os.environ[e]
+
+        if "LIBPYTHON_LOC" not in self.env:
+            self.env["LIBPYTHON_LOC"] = cocotb._vendor.find_libpython.find_libpython()
+
+        self.env["PATH"] += os.pathsep + cocotb.config.libs_dir
+
+        self.env["PYTHONPATH"] = os.pathsep.join(sys.path)
+        for path in self.python_search:
+            self.env["PYTHONPATH"] += os.pathsep + str(path)
+
+        prefix = sysconfig.get_config_var("prefix")
+        if prefix is not None:
+            self.env["PYTHONHOME"] = prefix
+
+        self.env["TOPLEVEL"] = self.sim_toplevel
+        self.env["MODULE"] = self.module
+
+        if not os.path.exists(self.build_dir):
+            os.makedirs(self.build_dir)
+
+    @abc.abstractmethod
+    def build_command(self) -> Sequence[Command]:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def test_command(self) -> Sequence[Command]:
+        raise NotImplementedError()
+
+    def build(
+        self,
+        library_name: str = "work",
+        verilog_sources: Sequence[PathLike] = [],
+        vhdl_sources: Sequence[PathLike] = [],
+        includes: Sequence[PathLike] = [],
+        defines: Sequence[str] = [],
+        parameters: Mapping[str, object] = {},
+        extra_args: Sequence[str] = [],
+        toplevel: Optional[str] = None,
+        always: bool = False,
+        build_dir: PathLike = "sim_build",
+        work_dir: Optional[PathLike] = None,
+    ):
+        self.init_dir(build_dir, work_dir)
+        os.makedirs(self.build_dir, exist_ok=True)
+
+        # note: to avoid mutating argument defaults, we ensure that no value
+        # is written without a copy. This is much more concise and leads to
+        # a better docstring than using `None` as a default in the parameters
+        # list.
+        self.library_name = library_name
+        self.verilog_sources = get_abs_paths(verilog_sources)
+        self.vhdl_sources = get_abs_paths(vhdl_sources)
+        self.includes = get_abs_paths(includes)
+        self.defines = list(defines)
+        self.parameters = dict(parameters)
+        self.compile_args = list(extra_args)
+        self.always = always
+        self.hdl_toplevel = toplevel
+
+        for e in os.environ:
+            self.env[e] = os.environ[e]
+
+        cmds = self.build_command()
+        self.execute(cmds)
+
+    def test(
+        self,
+        py_module: Union[str, Sequence[str]],
+        toplevel: str,
+        toplevel_lang: Optional[str] = None,
+        testcase: Optional[str] = None,
+        seed: Optional[Union[str, int]] = None,
+        python_search: Sequence[PathLike] = [],
+        extra_args: Sequence[str] = [],
+        plus_args: Sequence[str] = [],
+        extra_env: Mapping[str, str] = {},
+        waves: Optional[bool] = None,
+        gui: Optional[bool] = None,
+        build_dir: PathLike = "sim_build",
```

Comment:
@ktbarrett 
Can we make `build_dir` optional here and in this case use `build_dir` from `build`?